fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java

```diff
@@ -0,0 +1,302 @@
+                initAllStructure(conjunct.getChild(index), slotEqSlotPredicatesList,
+                        slotEqSlotPredicatesSet, slotToLiteralPredicatesList,
+                        slotToLiteralPredicatesSet, isNullPredicatesList, isNullPredicatesSet);
+            }
+        }
+        if (conjunct instanceof BinaryPredicate) {
+            Pair<Expr, Expr> pair = new Pair<>(conjunct.getChild(0), conjunct.getChild(1));
+            Pair<Expr, Expr> eqPair = new Pair<>(conjunct.getChild(1), conjunct.getChild(0));
+            if (conjunct.getChild(0) instanceof SlotRef
+                    && conjunct.getChild(1) instanceof LiteralExpr) {
+                if (!slotToLiteralPredicatesSet.contains(pair)
+                    && !slotToLiteralPredicatesSet.contains(eqPair)) {
+                    slotToLiteralPredicatesSet.add(pair);
+                    slotToLiteralPredicatesList.add(conjunct);
+                }
+            } else if (conjunct.getChild(0) instanceof SlotRef
+                    && conjunct.getChild(1) instanceof SlotRef) {
+                if (!slotEqSlotPredicatesSet.contains(pair)
+                    && !slotEqSlotPredicatesSet.contains(eqPair)) {
+                    slotEqSlotPredicatesSet.add(pair);
+                    slotEqSlotPredicatesList.add(conjunct);
+                }
+            }
+        } else if (conjunct instanceof IsNullPredicate) {
+            if (!isNullPredicatesSet.contains(conjunct.getChild(0))) {
+                isNullPredicatesSet.add(conjunct.getChild(0));
+                isNullPredicatesList.add(conjunct);
+            }
+        }
+    }
+
+    private void inferSlotPredicates(ArrayList<Expr> slotEqSlotPredicatesList,
+                                    Set<Pair<Expr, Expr>> slotToLiteralPredicatesSet,
+                                    ArrayList<Expr> slotToLiteralPredicatesList,
+                                    ArrayList<Expr> newPredicatesList,
+                                    Analyzer analyzer) {
+        ArrayList<Expr> tmpSlotToLiteralPredicatesList = new ArrayList<>();
+        for (Expr slotToLiteral : slotToLiteralPredicatesList) {
+            buildNewBinaryPredicate(slotToLiteral, slotEqSlotPredicatesList, slotToLiteralPredicatesSet,
+                    newPredicatesList, tmpSlotToLiteralPredicatesList, analyzer);
+        }
+
+        int tmpSize = 0;
+        do {
+            int startSize = tmpSlotToLiteralPredicatesList.size();
+
+            for (int index = 0; index < tmpSlotToLiteralPredicatesList.size(); ++index) {
+                Expr slotToLiteral = tmpSlotToLiteralPredicatesList.get(index);
+                buildNewBinaryPredicate(slotToLiteral, slotEqSlotPredicatesList, slotToLiteralPredicatesSet,
+                        newPredicatesList, tmpSlotToLiteralPredicatesList, analyzer);
+            }
+
+            for (int index = 0; index < newPredicatesList.size(); ++index) {
+                Expr slotToLiteral = newPredicatesList.get(index);
```

Comment:
Although this writing is functionally correct, there may be a risk of infinite loop if the logic is wrong.
Is it better to derive all the expressions of a single slotToLiteral predicate at once?